processing/src/main/java/org/apache/druid/segment/join/JoinFilterAnalyzer.java

```diff
@@ -0,0 +1,743 @@
+        IndexedTable.Reader reader = indexedTable.columnReader(correlatedColumnPosition);
+        IntList rowIndex = index.find(filterValue);
+        List<String> correlatedValues = new ArrayList<>();
+        for (int i = 0; i < rowIndex.size(); i++) {
+          int rowNum = rowIndex.getInt(i);
+          correlatedValues.add(reader.read(rowNum).toString());
+        }
+        return correlatedValues;
+      } else {
+        IndexedTable.Reader dimNameReader = indexedTable.columnReader(filterColumnPosition);
+        IndexedTable.Reader correlatedColumnReader = indexedTable.columnReader(correlatedColumnPosition);
+        Set<String> correlatedValueSet = new HashSet<>();
+        for (int i = 0; i < indexedTable.numRows(); i++) {
+          if (filterValue.equals(dimNameReader.read(i).toString())) {
+            correlatedValueSet.add(correlatedColumnReader.read(i).toString());
+          }
+        }
+
+        return new ArrayList<>(correlatedValueSet);
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * For all RHS columns that appear in the join's equiconditions, correlate them with base table columns if possible.
+   *
+   * @param adapter              The adapter for the join. Used to determine if a column is a base table column.
+   * @param tablePrefix          Prefix for a join table
+   * @param clauseForTablePrefix Joinable clause for the prefix
+   * @param equiconditions       Map of equiconditions, keyed by the right hand columns
+   *
+   * @return A list of correlatation analyses for the equicondition RHS columns that reside in the table associated with
+   * the tablePrefix
+   */
+  @Nullable
+  private static List<JoinFilterColumnCorrelationAnalysis> findCorrelatedBaseTableColumns(
+      HashJoinSegmentStorageAdapter adapter,
+      String tablePrefix,
+      JoinableClause clauseForTablePrefix,
+      Map<String, Expr> equiconditions
+  )
+  {
+    JoinConditionAnalysis jca = clauseForTablePrefix.getCondition();
+
+    List<String> rhsColumns = new ArrayList<>();
+    for (Equality eq : jca.getEquiConditions()) {
+      rhsColumns.add(tablePrefix + eq.getRightColumn());
+    }
+
+    List<JoinFilterColumnCorrelationAnalysis> correlations = new ArrayList<>();
+
+    for (String rhsColumn : rhsColumns) {
+      List<String> correlatedBaseColumns = new ArrayList<>();
+      List<Expr> correlatedBaseExpressions = new ArrayList<>();
+      boolean terminate = false;
+
+      String findMappingFor = rhsColumn;
+      while (!terminate) {
+        Expr lhs = equiconditions.get(findMappingFor);
+        if (lhs == null) {
+          break;
+        }
+        String identifier = lhs.getBindingIfIdentifier();
+        if (identifier == null) {
+          // We push down if the function only requires base table columns
+          Expr.BindingDetails bindingDetails = lhs.analyzeInputs();
+          Set<String> requiredBindings = bindingDetails.getRequiredBindings();
+          for (String requiredBinding : requiredBindings) {
```

Comment:
small suggestion: `!requiredBindings.stream().allMatch( blah blah blah )` may be more readable

Reply:
Changed to the suggested stream usage